cmd/notation/blob/inspect.go

```diff
@@ -0,0 +1,120 @@
+// limitations under the License.
+
+package blob
+
+import (
+	"errors"
+	"fmt"
+	"github.com/notaryproject/notation-core-go/signature"
+	"github.com/notaryproject/notation-go/plugin/proto"
+	"github.com/notaryproject/notation/cmd/notation/internal/outputs"
+	"github.com/notaryproject/notation/internal/cmd"
+	"github.com/notaryproject/notation/internal/envelope"
+	"github.com/spf13/cobra"
+	"os"
+	"path/filepath"
+)
+
+type blobInspectOpts struct {
+	cmd.LoggingFlagOpts
+	cmd.SignerFlagOpts
+	signaturePath string
+	outputFormat  string
+	mediaType     string
+}
+
+func inspectCommand(opts *blobInspectOpts) *cobra.Command {
+	if opts == nil {
+		opts = &blobInspectOpts{}
+	}
+	longMessage := `Inspect signature associated with the signed blob.
+
+Example - Inspect BLOB signature:
+  notation blob inspect <signature_path>
+
+Example - Inspect BLOB signature and output as JSON:
+  notation blob inspect --output json <signature_path>
+`
+
+	command := &cobra.Command{
+		Use:   "blob inspect [signaturePath]",
+		Short: "Inspect signature associated with the signed BLOB",
+		Long:  longMessage,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("missing signature path to the artifact: use `notation blob inspect --help` to see what parameters are required")
+			}
+			opts.signaturePath = args[0]
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runBlobInspect(opts)
+		},
+	}
+
+	opts.LoggingFlagOpts.ApplyFlags(command.Flags())
+	cmd.SetPflagOutput(command.Flags(), &opts.outputFormat, cmd.PflagOutputUsage)
+	return command
+}
+
+func runBlobInspect(opts *blobInspectOpts) error {
+	if opts.outputFormat != cmd.OutputJSON && opts.outputFormat != cmd.OutputPlaintext {
+		return fmt.Errorf("unrecognized output format %s", opts.outputFormat)
+	}
+
+	// initialize
+	mediaType, err := envelope.GetEnvelopeMediaType(filepath.Ext(opts.signaturePath))
+	if !(mediaType == "jws.MediaTypeEnvelope" || mediaType == "cose.MediaTypeEnvelope") {
+		return err
+	}
+	contents, err := os.ReadFile(opts.signaturePath)
+	if err != nil {
+		return err
+	}
+	output := outputs.InspectOutput{MediaType: mediaType, Signatures: []outputs.SignatureOutput{}}
+	sigEnvelope, err := signature.ParseEnvelope(mediaType, contents)
+	if err != nil {
+		return err
+	}
+	envelopeContent, err := sigEnvelope.Content()
+	if err != nil {
+		return err
+	}
+	signedArtifactDesc, err := envelope.DescriptorFromSignaturePayload(&envelopeContent.Payload)
+	if err != nil {
+		return err
+	}
+	signatureAlgorithm, err := proto.EncodeSigningAlgorithm(envelopeContent.SignerInfo.SignatureAlgorithm)
+	if err != nil {
+		return err
+	}
+
+	sig := outputs.SignatureOutput{
+		MediaType:             mediaType,
+		SignatureAlgorithm:    string(signatureAlgorithm),
+		SignedAttributes:      outputs.GetSignedAttributes(opts.outputFormat, envelopeContent),
+		UserDefinedAttributes: signedArtifactDesc.Annotations,
```

Comment:
When we are already passing SignedArtifact, why do we need to pass `UserDefinedAttributes` ?

Also looks like code from line 90 to 110 is repeated in `cmd/notation/inspect.go` can we just take `sigEnvelope` in `outputs.SignatureOutput` and do all this computation inside `PrintOutput` function?


Also for, `logSkippedSignature` functionality you can pass a function to `PrintOutput` method that either consumes the error and prints the warning or re-throws the error causing `PrintOutput` to return error.

Reply:
Code duplication has been removed by using a common function